@W-16567765 Do Not Track UI

jeremy-jung1 opened the pull request:
Add the modal for DNT in Template Retail React App

# Description

<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relevant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [x] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [x] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- (step1) Go to https://runtime-admin-n16.mobify-storefront.com/ and see that the consent form shows up
- (step2) Click outside the modal to see that the modal disappears. Check Developer Tools -> Application -> Cookies to see that dw_dnt is created with its value to 1
- (step3) Delete the cookie and refresh to see that the modal reappears. Click on 'Accept'. Check that dw_dnt with value set to 0 is created in cookies.
- (step4) Delete the cookie and click on another page. See that modal reappears. Click on 'Decline' to see that dw_dnt is created with 1. 
- (step5) Delete the cookie and refresh. Click on the 'X' at the modal to see that dw_dnt with 0 is created. 
- (step6) Go to another page and see that the modal does not pop up again


# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [x] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

jeremy-jung1 committed:
Make UI

jeremy-jung1 committed:
Translations and some CSS

jeremy-jung1 committed:
Update CHANGELOG.md

jeremy-jung1 committed:
Adding mobile view

alexvuong commented on packages/template-retail-react-app/app/hooks/use-dnt-modal.js:
```diff
@@ -0,0 +1,156 @@
+
+    const buttons = (
+        <>
+            <Button
```
Can you add aria-labels on these button? Something like aria-labels={intl.formatMessage({id: 'dnw_modal.assistive_msg.defline', defaultMessage: "Decline Do not Track"})}

alexvuong commented:
So we are not showing overlay background when modal is showing, and it disappears when you happen to click outside of the modal. This potential causes confusion. I think we should add an overlayer to the modal to highlight the modal. 
![image](https://github.com/user-attachments/assets/b88dc7a3-2bd9-4d65-aab2-e17882fab8e3)
SFRA: 
![image](https://github.com/user-attachments/assets/203f4889-3d21-448a-9cf8-5f673b4e0c33)

alexvuong commented:
How can we test this modal on Hybrid flow?

jeremy-jung1 committed:
Merge branch 'dnt-feature-branch' into W-16567765-dnt-ui

jeremy-jung1 committed:
Make content behind modal interactive

alexvuong commented on packages/template-retail-react-app/app/hooks/use-dnt-modal.js:
```diff
@@ -0,0 +1,153 @@
+    )
+
+    return (
+        isOpen && <Box
```
Any reason we are moving away from Chakra Modal?

jeremy-jung1 replied:
Chakra UI doesn't allow for the content behind the modal to be interactive. 
> Quoting from the docs: [content behind a modal dialog is inert, meaning that users cannot interact with it](https://v2.chakra-ui.com/docs/components/modal)

The most the Chakra UI can allow for the background content is to be able to scroll or for some screen readers to read content. I looked for and tried supported props that could be passed into the Modal that could allow for interaction, but there wasn't any that could fix this issue.

jeremy-jung1 committed:
linting

jeremy-jung1 committed:
Update use-dnt-modal.js

jeremy-jung1 commented:
@alexvuong 
Re: background of the modal
After feedback from Trevor to make the content behind the modal interactive, I addressed your feedback in tandum about the confusion regarding the background of the modal by making it all interactive (meaning clicking out of the modal does not close the form but instead users can click on the website). Users will find that the form disappears if they press Accept, Decline, the X button, or have created `dw_dnt` in SFRA.
Re: hybrid testing
I tested this modal in Hybrid flow using the `bjnl_prd` environment. I found that from SFRA->PWA the cookie is successfully transitioned to PWA, and the form does not reappear. From PWA -> SFRA, there is a weird behavior in SFRA that overwrites the cookie. This is something that I'll look into, but is out of scope for this PR that is only UI in PWA.

Tomorrow morning, I'll deploy the hybrid instance again so others can see it

jeremy-jung1 committed:
linting

jeremy-jung1 committed:
Updates on message

alexvuong commented on packages/template-retail-react-app/app/hooks/use-dnt-modal.js:
```diff
@@ -0,0 +1,153 @@
+                    onClose()
+                }}
+                aria-label={formatMessage({
+                    id: 'dnt_modal.button.decline.assistive_msg',
```
```suggestion
                    id: 'dnt_modal.button.assistive_msg.decline',
```

alexvuong commented on packages/template-retail-react-app/app/hooks/use-dnt-modal.js:
```diff
@@ -0,0 +1,153 @@
+                })}
+                width="100%"
+            >
+                <FormattedMessage defaultMessage="Decline" id="dnt_modal.decline" />
```
```suggestion
                <FormattedMessage defaultMessage="Decline" id="dnt_modal.button.decline" />
```

alexvuong commented on packages/template-retail-react-app/CHANGELOG.md:
```diff
@@ -9,7 +9,7 @@
 - PDP / PLP: Render non HTTP 404 erros [#2003](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2003)
 - Error page: Render home page when clicking nav icon [#2003](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2003)
-
+- DNT UI: [#2017](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2017)
```
```suggestion
- DNT UI: [#2017](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2017)

```

alexvuong commented on packages/template-retail-react-app/app/components/_app/index.test.js:
```diff
@@ -61,15 +62,20 @@ describe('App', () => {
-        expect(screen.getByRole('main')).toBeInTheDocument()
-        expect(screen.getByText('Any children here')).toBeInTheDocument()
+        const closeButton = screen.getByLabelText('Close dnt form')
+        const user = userEvent.setup()
```
user is already included in the renderWithProvider (see [here](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/develop/packages/template-retail-react-app/app/utils/test-utils.js#L197)). You can get the user object like this, you don't need to call user.setup() 
```
const {user} = nrenderWithProviders(
            <App targetLocale={DEFAULT_LOCALE} defaultLocale={DEFAULT_LOCALE} messages={messages}>
                <p>Any children here</p>
            </App>
        )
```

alexvuong commented on packages/template-retail-react-app/app/components/_app/index.test.js:
```diff
@@ -61,15 +62,20 @@ describe('App', () => {
     }
 
-    test('App component is rendered appropriately', () => {
+    test('App component is rendered appropriately', async () => {
```
```suggestion
    test('User can select DNT options when App component is rendered with DNT Modal', async () => {
```

alexvuong commented:
@jeremy-jung1, thanks for addressing the feedback. A few questions:
- Why are we not using Chakra Modal for this feature? I'd recommend to use it since it comes with a11y OOTB, we don't have to think much if we can utilize the component
- Since the DNT Modal is not focus on render, and it is at the bottom of the page. From a11y POV,  how does a visually impaired person be aware of it when they land on the page? It would take a lot of tabs to be able to get into the modal when you are using Tab key

jeremy-jung1 committed:
Close form when back button is pressed from SFRA

vmarta commented:
> Since the DNT Modal is not focus on render, and it is at the bottom of the page. From a11y POV, how does a visually impaired person be aware of it when they land on the page? It would take a lot of tabs to be able to get into the modal when you are using Tab key

Perhaps we can simply place the modal near the _top_ of the document's `<body>`. So that the screen reader will reach it quickly. Visually it can still float above the page, at the bottom of the viewport.

vmarta commented:
@jeremy-jung1 let's make sure that the modal does not cover the page's footer. For example, currently we can't click on the country/locale selector and also some of the footer links like the Privacy Policy.

vmarta commented:
Aside: since the UI element we wanted should not block users from interacting with the page content, I think we should not call it a Modal. A modal means that the users should focus and interact with the modal first, before they can go back to interacting with the page.

What we wanted sounds like a _notification_. It shows up but does not force you to interact with it immediately.

jeremy-jung1 committed:
Some updates based on feedback

jeremy-jung1 committed:
Updates after feedback

jeremy-jung1 committed:
Update modal wording to notification

jeremy-jung1 committed:
linting

jeremy-jung1 committed:
Update more wording

jeremy-jung1 committed:
Increase bundlesize

jeremy-jung1 committed:
Merge branch 'develop' into W-16567765-dnt-ui

jeremy-jung1 committed:
Merge branch 'dnt-feature-branch' into W-16567765-dnt-ui

jeremy-jung1 committed:
Remove unused import

jeremy-jung1 commented:
@alexvuong @vmarta 
I [asked an accessibility engineer in #tech-prod-ally that the form being the last element to be tabbed to in the page works](https://salesforce-internal.slack.com/archives/C022Q6STSHG/p1727899778947999?thread_ts=1727717075.422099&cid=C022Q6STSHG). The reason I asked was because I found that for accessibility and maintenance reasons modifying the ordering of the components ( i.e. through tabIndexing) seemed to be discouraged from what I read. Furthermore, chakra UI creates a Portal for Modals and other popup components, which [deliberately puts the contents at the *end* of document.body](https://v2.chakra-ui.com/docs/components/modal/usage#keyboard-and-focus-management) for accessibility purposes, making putting the element near the top of the body potentially involve circumventing Chakra UI's accessibility features. I also haven't found a straightforward way to bump the natural order of the modal in the tabbing sequence. Using `useRef` and `focus` would only allow for the form to be focused immediately but still keeps its ordering at the bottom of the body. 

Should the notification still be one of the first things tab-able in the page?

alexvuong commented on packages/template-retail-react-app/app/hooks/use-dnt-notification.js:
```diff
@@ -0,0 +1,176 @@
+
+/**
+ *
+ * @returns {Object} - Object props to be spread on to the DntNotification component
```
Nit
```suggestion
 * @returns {Object} - Object props to be passed into  DntNotification component
```

jeremy-jung1 committed:
Wording from modal to notification

alexvuong commented on packages/template-retail-react-app/app/hooks/use-dnt-notification.test.js:
```diff
@@ -0,0 +1,82 @@
+    jest.resetModules()
+})
+
+test('Modal gives the expected text', async () => {
```
Nit
```suggestion
test('Modal renders properly', async () => {
```

alexvuong commented on packages/template-retail-react-app/app/hooks/use-dnt-notification.test.js:
```diff
@@ -0,0 +1,82 @@
+test('Modal gives the expected text', async () => {
+    renderWithProviders(<MockedComponent />)
+    await waitFor(() => {
+        expect(screen.getByText(/Tracking Consent/i)).toBeInTheDocument()
```
Small ask, would you please check to see if aria-label is also present in the DOM?

alexvuong commented on packages/template-retail-react-app/app/hooks/use-dnt-notification.test.js:
```diff
@@ -0,0 +1,82 @@
+})
+
+test('Clicking out of modal does setDNT(null)', async () => {
+    const user = userEvent.setup()
```
We don't need to re-setup user here

alexvuong commented on packages/template-retail-react-app/app/hooks/use-dnt-notification.test.js:
```diff
@@ -0,0 +1,82 @@
+})
+
+test('Clicking Accept does setDNT(false)', async () => {
+    const user = userEvent.setup()
```
Same with above

alexvuong commented on packages/template-retail-react-app/app/hooks/use-dnt-notification.test.js:
```diff
@@ -0,0 +1,82 @@
+})
+
+test('Clicking Decline does setDNT(true)', async () => {
+    const user = userEvent.setup()
```
Same with above

alexvuong reviewed (approved):
Left some small and non-blocking comments. Thanks for working on this feature. Great work 👍

jeremy-jung1 committed:
await updateDNT

vmarta commented:
> Should the notification still be one of the first things tab-able in the page?

Now that we're using Chakra Modal component, let's just follow its lead. Like you said, it uses the Portal, so yeah I'm ok with the notification being near the end of the DOM.

vmarta commented on packages/template-retail-react-app/app/hooks/use-dnt-notification.js:
```diff
@@ -0,0 +1,176 @@
+    }, [location, dntStatus])
+
+    const onCloseNotification = async () => {
+        await updateDNT(null)
+        onClose()
```
```suggestion
        updateDNT(null)
        onClose()
```

There are several calls to `updateDNT` that are followed by closing the notification 🤔  Yes, `updateDNT` is an async operation, but I'm not sure if awaiting it will be good for the user experience.

Usually the act of clicking a button should result in immediate feedback. But in this case, it's possible to see a delay, while we wait for `updateDNT` to finish. If it's not crucial to await, I'd prefer to not await and let `updateDNT` run in the background.

jeremy-jung1 replied:
@alexvuong recommended we await the updateDNT in case another reauthentication happens around the same time, at which point if we don't await this call then that second call could conflict and overwrite the token's DNT and we could lose the update here.

vmarta reviewed (approved):
One minor comment. I'll leave it up to you. Other than that, this is good 👍 Thanks for iterating and addressing all the feedback!

vmarta commented on packages/template-retail-react-app/app/hooks/use-dnt-notification.js:
```diff
@@ -0,0 +1,176 @@
+    )
+
+    return (
+        <Modal
+            size="sm"
+            data-testid="sf-dnt-notification"
+            blockScrollOnMount={false}
+            closeOnOverlayClick={false}
+            trapFocus={false}
```
Forgot to mention that it's cool that we can customize the Modal to behave more like a notification and the page content is still interactive 👍

jeremy-jung1 committed:
Apply feedback

jeremy-jung1 committed:
Linting

jeremy-jung1 committed:
Show form again when cookie conflict with token

jeremy-jung1 committed:
Make updateDNT async

jeremy-jung1 committed:
Update CHANGELOG.md

jeremy-jung1 committed:
Update tests

jeremy-jung1 committed:
Handle token inexistence

jeremy-jung1 commented on packages/commerce-sdk-react/src/auth/index.ts:
```diff
@@ -290,9 +290,15 @@ class Auth {
+        const accessToken = this.getAccessToken()
+        let hasNoConflictWithToken = true
+        if (accessToken) {
+            const {dnt} = this.parseSlasJWT(accessToken)
```
Conditional to handle case when the retrieval of the DNT state occurs before a token is stored in the browser

jeremy-jung1 commented on packages/commerce-sdk-react/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## Version on which DNT SDK will go live (TODO: Update/cleanup when merging to develop)
+- Added useDNT hook to commerce-sdk-react [#1989](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1989) Feature branch
```
Modified CHANGELOG to pass the CI but will squash the PRs when releasing

vmarta commented on packages/commerce-sdk-react/src/auth/index.ts:
```diff
@@ -290,9 +290,15 @@ class Auth {
+            const {dnt} = this.parseSlasJWT(accessToken)
+            hasNoConflictWithToken = dnt === dntCookieVal
+        }
+        if ((dntCookieVal !== '1' && dntCookieVal !== '0') || !hasNoConflictWithToken) {
```
Minor: sometimes it's easier to parse/understand if the variable name does not have a negative in it. So with the `!` it's like dealing with a double negatives.
 
```suggestion
        if ((dntCookieVal !== '1' && dntCookieVal !== '0') || !isInSync) {
```

jeremy-jung1 committed:
Remove double negative
